### PR TITLE
Add native responses

### PR DESCRIPTION
**Pull request by enyst**
Native responses API, take two (or four 😅)
- responses() stateless, with its reasoning, sent back and shown off
- some preparation for stateful, for another PR

This PR proposes an approach where:
- the Agent is aware of the two completion() and responses()
- but I've tried to keep its changes minimal, refactoring out low-level details of parsing LLM responses, and LLM tool calls
- llm.py implements both (but we will probably change this)
  - the code is not fully clean atm, but generally I've tried to sort of keep it small and "symmetrical", so it's easy to refactor out
- message.py, small new interface:
  - `from_llm_chat_message`
  - `from_llm_responses_output`
  - `to_chat_dict`
  - `to_responses_dict`

---
**Agent Server images for this PR**

• **GHCR package:** https://github.com/All-Hands-AI/agent-sdk/pkgs/container/agent-server

**Variants & Base Images**
| Variant | Base Image | Docs / Tags |
|---|---|---|
| golang | `golang:1.21-bookworm` | [Link](https://hub.docker.com/_/golang:1.21-bookworm) |
| java | `eclipse-temurin:17-jdk` | [Link](https://hub.docker.com/_/eclipse-temurin:17-jdk) |
| python | `nikolaik/python-nodejs:python3.12-nodejs22` | [Link](https://hub.docker.com/_/nikolaik/python-nodejs:python3.12-nodejs22) |

**Pull (multi-arch manifest)**
```bash
docker pull ghcr.io/all-hands-ai/agent-server:eb429ca-python
```

**Run**
```bash
docker run -it --rm \
  -p 8000:8000 \
  --name agent-server-eb429ca-python \
  ghcr.io/all-hands-ai/agent-server:eb429ca-python
```

**All tags pushed for this build**
```
ghcr.io/all-hands-ai/agent-server:eb429ca-golang
ghcr.io/all-hands-ai/agent-server:v1.0.0_golang_tag_1.21-bookworm_golang
ghcr.io/all-hands-ai/agent-server:eb429ca-java
ghcr.io/all-hands-ai/agent-server:v1.0.0_eclipse-temurin_tag_17-jdk_java
ghcr.io/all-hands-ai/agent-server:eb429ca-python
ghcr.io/all-hands-ai/agent-server:v1.0.0_nikolaik_s_python-nodejs_tag_python3.12-nodejs22_python
```

_The `eb429ca` tag is a multi-arch manifest (amd64/arm64); your client pulls the right arch automatically._

### CODE REVIEW

**Comment by enyst** [comment]
@openhands Extract NonExecutableActionEvent from this PR, make a *new branch* from main, and apply it on it. Then make a PR.

Understand well what role it has, and understand that on main branch we don't have Responses path. Make sure that NonExecutableActionEvent fulfills its role correctly on Completions path. Add unit tests to that new PR.

**Comment by enyst** [comment]
This works again! GPT-5-mini thinks for a while when deleting a file
(this also shows the other issue)
image

**Comment by enyst** [comment]
This works! View and send back reasoning on stateless variant, and it cleared up some of the code. At a price (llm.py is still a bit unhappy, and telemetry.py will need another pass), just IMHO it's mostly easier to read and follow the execution path for one API or the other.

I kinda like that the Agent only chooses, all the rest of its code is now without `ifs` and without litellm/completions code.

`message.py` saw a heavier lift, @xingyaoww I'd love to know what you think, I admit I like to look at a random screen of code and know exactly what belongs where. e.g. Tiny lost bits like `_add_tool_call_keys` were not obvious (once we have responses, are they executed on responses?), so this PR took them out. I still want to reintroduce the types for parameters, but I feel it's probably best separate.

The agent is working on porting tests.

I think llm.py needs a heavier refactoring..., I'm thinking to make a facade for normalize_kwargs and format_messages_for(), and take them out; not sure it'd reduce redundancies, but more for readability and more reliance on types. 🤔

This PR doesn't do a conversion; no conversion the other way around either
- this is [Option 2 with some bits of 3.1](https://github.com/All-Hands-AI/agent-sdk/issues/138#issuecomment-3293323214)

**Review by xingyaoww — COMMENTED** [review]
Did a quick skim thourgh! I actually think the structure LGTM and make sense!

Happy to take another look when we merge that NonExecutableActionEvent and when tests are in and passed!

🙏

File: .gitignore [hunk 1]
```diff
@@ -203,3 +203,5 @@ cache
 /workspace/
 openapi.json
 .client/
+# Git worktrees
+.worktrees/
```
Comment (enyst) [inline .gitignore]:
```suggestion
```

File: examples/01_hello_world.py [hunk 2]
```diff
@@ -69,7 +69,8 @@ def conversation_callback(event: Event):
 )
 
 conversation.send_message(
-    "Read the current repo and write 3 facts about the project into FACTS.txt."
+    "Read the current repo README.md and write 1 line on top"
+    " of the file, using a tool call with insert line. You must use insert line."
```
Comment (enyst) [inline examples/01_hello_world.py]:
```suggestion
    "Read the current repo and write 3 facts about the project into FACTS.txt."
```

File: examples/26_responses_reasoning.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+    print("=" * 100)
+    print("Conversation finished. Got the following LLM messages:")
+    for i, message in enumerate(llm_messages):
+        ms = str(message)
+        print(f"Message {i}: {ms[:200]}{'...' if len(ms) > 200 else ''}")
+
+
+def main():
+    # Prefer proxy credentials commonly used in this repo's examples
```
Comment (enyst) [inline examples/26_responses_reasoning.py]:
```suggestion
```

**Comment by enyst** [comment]
@neubig If you may want to see this PR, it's the newest (rebuilt) for Responses API, and the one I'd love to try to get in.

**Comment by enyst** [comment]
Adding design docs for LLM refactor (readability, streaming/async, stateful Responses) and model-info/capabilities init.

Docs:
- docs/design/llm-refactor.md
- docs/design/llm-model-info-and-caps.md

Summary:
- Thin LLM facade; extract formatters, select_options, tool strategy (native vs mock), transport, invokers.
- Cache model_info resolution; derive token limits and FC capability; group Anthropic logic (cache markers, extended thinking headers, practical caps).
- No behavior changes; groundwork for streaming/async/stateful Responses.

Open to feedback on boundaries and naming before starting extractions.

File: openhands/sdk/llm/message.py [hunk 4]
```diff
@@ -181,34 +248,47 @@ def _coerce_content(cls, v: Any) -> Sequence[TextContent | ImageContent] | Any:
...
+        - Assistant tool call turn: role == "assistant" and self.tool_calls
+        - Tool result turn: role == "tool" and self.tool_call_id (with name)
         """
         if not self.force_string_serializer and (
             self.cache_enabled or self.vision_enabled or self.function_calling_enabled
         ):
             message_dict = self._list_serializer()
         else:
-            # some providers, like HF and Groq/llama, don't support a list here, but a
-            # single string
```
Comment (enyst) [inline openhands/sdk/llm/message.py:198]:
```Suggestion
            # some providers, like HF and Groq/llama, don't support a list here, but a
            # single string
```

File: openhands/sdk/llm/message.py [hunk 5]
```diff
@@ -181,34 +248,47 @@ def _coerce_content(cls, v: Any) -> Sequence[TextContent | ImageContent] | Any:
...
             message_dict = self._string_serializer()
```
Comment (enyst) [inline openhands/sdk/llm/message.py:265]:
```Suggestion
            # some providers, like HF and Groq/llama, don't support a list here, but a
            # single string
            message_dict = self._string_serializer()
```

**Comment by enyst** [comment]
OpenHands-GPT-5 here. Concise summary of test coverage for the Responses API work:

- Agent.step routing/gating
  - Override gate: uses_responses_api() True -> responses(); False -> completion(); assistant MessageEvent emitted
  - Model gate: gpt-5-mini-2025-08-07 -> responses(); gpt-4o-mini -> completion()
- Responses API parsing & kwargs
  - from_llm_responses_output: assistant text join, function call normalization (fc_ prefix), reasoning summary mapping
  - to_responses_dict: fc_ prefix and arg stringification
  - _normalize_responses_kwargs: temperature=1.0 for responses; include adds reasoning.encrypted_content when enabled; store defaults False; max_output_tokens passthrough
  - End-to-end responses() (patched) verifies Telemetry records a usage entry
- Message & tool coverage
  - from_llm_chat_message edge cases; _coerce_content/content_to_str; system to_responses_value
  - ToolBase.to_responses_tool: security_risk gating tests
- Status
  - Pre-commit clean (ruff, pyright); full pytest green: 1507 passed, 35 warnings
- Follow-ups (optional)
  - Telemetry tests for cached_tokens and reasoning_tokens mapping; audit remaining openai vs litellm imports

**Comment by neubig** [comment]
Thanks for this! I'll try to take a look sometime in these two days!

**Comment by enyst** [comment]
> Did a quick skim thourgh! I actually think the structure LGTM and make sense! (and agree that we might need to do a good refactor afterwards)
> 
> Happy to take another look when we merge that NonExecutableActionEvent and when tests are in and passed!
> 
> 🙏

Merged NEA and tests!

File: openhands/sdk/tool/tool.py [hunk 6]
```diff
@@ -309,6 +313,36 @@ def to_openai_tool(
...
+        action_type_with_risk = _create_action_type_with_risk(action_type)
+
+        add_security_risk_prediction = add_security_risk_prediction and (
+            self.annotations is None or (not self.annotations.readOnlyHint)
+        )
+        schema = (
+            action_type_with_risk.to_mcp_schema()
+            if add_security_risk_prediction
+            else action_type.to_mcp_schema()
+        )
```
Comment (xingyaoww) [inline openhands/sdk/tool/tool.py]: nit: we should able to duduplicate this with to_openai_tool

File: openhands/sdk/llm/message.py [hunk 7]
```diff
@@ -198,17 +264,33 @@ def to_llm_dict(self) -> dict[str, Any]:
...
+        if self.role == "tool" and self.tool_call_id is not None:
+            assert self.name is not None, (
+                "name is required when tool_call_id is not None"
+            )
+            message_dict["tool_call_id"] = self.tool_call_id
+            message_dict["name"] = self.name
+
         return message_dict
 
+    # Backward-compatible alias
```
Comment (xingyaoww) [inline openhands/sdk/llm/message.py:281]: Let's clean clean this up in future PRs? We can just remove this?

File: openhands/sdk/llm/message.py [hunk 8]
```diff
@@ -332,11 +509,74 @@ def from_litellm_message(cls, message: LiteLLMMessage) -> "Message":
...
+        assistant_text = "\n".join(assistant_text_parts).strip()
+        return Message(
+            role="assistant",
+            content=[TextContent(text=assistant_text)] if assistant_text else [],
+            tool_calls=tool_calls or None,
+            responses_reasoning_item=responses_reasoning_item,
+        )
+
+    @classmethod
+    def from_litellm_message(cls, message: LiteLLMMessage) -> "Message":
```
Comment (xingyaoww) [inline openhands/sdk/llm/message.py:576]: Let's remove this too in follow up PR

**Review by xingyaoww — COMMENTED** [review]
I made some changes along the way but this PR LGTM! There's still some changes for simplification / clean up but we can leave for future PRs


One last issue i'm trying to figure out now: https://community.openai.com/t/schema-additionalproperties-must-be-false-when-strict-is-true/929996


    # Keep the tasks short for demo purposes
    ^^^^^^^^^^^^^^^^^^
  File "/Users/xingyaow/Projects/All-Hands-AI/openhands-v1-dev/agent-sdk.worktree/worktree2/openhands/sdk/conversation/impl/local_conversation.py", line 244, in run
    self.agent.step(self._state, on_event=self._on_event)
  File "/Users/xingyaow/Projects/All-Hands-AI/openhands-v1-dev/agent-sdk.worktree/worktree2/openhands/sdk/agent/agent.py", line 209, in step
    raise e
  File "/Users/xingyaow/Projects/All-Hands-AI/openhands-v1-dev/agent-sdk.worktree/worktree2/openhands/sdk/agent/agent.py", line 178, in step
    llm_response = self.llm.responses(
                   ^^^^^^^^^^^^^^^^^^^
  File "/Users/xingyaow/Projects/All-Hands-AI/openhands-v1-dev/agent-sdk.worktree/worktree2/openhands/sdk/llm/llm.py", line 593, in responses
    resp: ResponsesAPIResponse = _one_attempt()
                                 ^^^^^^^^^^^^^^
  File "/Users/xingyaow/Projects/All-Hands-AI/openhands-v1-dev/agent-sdk.worktree/worktree2/.venv/lib/python3.12/site-packages/tenacity/__init__.py", line 338, in wrapped_f
    return copy(f, *args, **kw)
           ^^^^^^^^^^^^^^^^^^^^
  File "/Users/xingyaow/Projects/All-Hands-AI/openhands-v1-dev/agent-sdk.worktree/worktree2/.venv/lib/python3.12/site-packages/tenacity/__init__.py", line 477, in __call__
    do = self.iter(retry_state=retry_state)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/xingyaow/Projects/All-Hands-AI/openhands-v1-dev/agent-sdk.worktree/worktree2/.venv/lib/python3.12/site-packages/tenacity/__init__.py", line 378, in iter
    result = action(retry_state)
             ^^^^^^^^^^^^^^^^^^^
  File "/Users/xingyaow/Projects/All-Hands-AI/openhands-v1-dev/agent-sdk.worktree/worktree2/.venv/lib/python3.12/site-packages/tenacity/__init__.py", line 400, in <lambda>
    self._add_action_func(lambda rs: rs.outcome.result())
                                     ^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.12/3.12.11/Frameworks/Python.framework/Versions/3.12/lib/python3.12/concurrent/futures/_base.py", line 449, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.12/3.12.11/Frameworks/Python.framework/Versions/3.12/lib/python3.12/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/Users/xingyaow/Projects/All-Hands-AI/openhands-v1-dev/agent-sdk.worktree/worktree2/.venv/lib/python3.12/site-packages/tenacity/__init__.py", line 480, in __call__
    result = fn(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^
  File "/Users/xingyaow/Projects/All-Hands-AI/openhands-v1-dev/agent-sdk.worktree/worktree2/openhands/sdk/llm/llm.py", line 569, in _one_attempt
    ret = litellm_responses(
          ^^^^^^^^^^^^^^^^^^
  File "/Users/xingyaow/Projects/All-Hands-AI/openhands-v1-dev/agent-sdk.worktree/worktree2/.venv/lib/python3.12/site-packages/litellm/utils.py", line 1343, in wrapper
    raise e
  File "/Users/xingyaow/Projects/All-Hands-AI/openhands-v1-dev/agent-sdk.worktree/worktree2/.venv/lib/python3.12/site-packages/litellm/utils.py", line 1218, in wrapper
    result = original_function(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/xingyaow/Projects/All-Hands-AI/openhands-v1-dev/agent-sdk.worktree/worktree2/.venv/lib/python3.12/site-packages/litellm/responses/main.py", line 655, in responses
    raise litellm.exception_type(
          ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/xingyaow/Projects/All-Hands-AI/openhands-v1-dev/agent-sdk.worktree/worktree2/.venv/lib/python3.12/site-packages/litellm/litellm_core_utils/exception_mapping_utils.py", line 2301, in exception_type
    raise e
  File "/Users/xingyaow/Projects/All-Hands-AI/openhands-v1-dev/agent-sdk.worktree/worktree2/.venv/lib/python3.12/site-packages/litellm/litellm_core_utils/exception_mapping_utils.py", line 391, in exception_type
    raise BadRequestError(
litellm.exceptions.BadRequestError: litellm.BadRequestError: OpenAIException - {"error":{"message":"litellm.BadRequestError: OpenAIException - {\n  \"error\": {\n    \"message\": \"Invalid schema for function 'execute_bash': In context=(), 'additionalProperties' is required to be supplied and to be false.\",\n    \"type\": \"invalid_request_error\",\n    \"param\": \"tools[0].parameters\",\n    \"code\": \"invalid_function_parameters\"\n  }\n}. Received Model Group=gpt-5-mini\nAvailable Model Group Fallbacks=None","type":null,"param":null,"code":"400"}}
/Users/xingyaow/Projects/All-Hands-AI/openhands-v1-dev/agent-sdk.worktree/worktree2/.venv/lib/python3.12/site-packages/litellm/llms/custom_httpx/async_client_cleanup.py:66: DeprecationWarning: There is no current event loop
  loop = asyncio.get_event_loop()

File: openhands/sdk/tool/tool.py (hunk 6 above)
Comment (xingyaoww) [reply openhands/sdk/tool/tool.py]: I did it in https://github.com/All-Hands-AI/agent-sdk/pull/622/commits/c9c5b38482daabdae357bf97c9c474d5b2bcb175

**Comment by enyst** [comment]
@xingyaoww If it helps, actually we had it unspecified most of the time
Trying now 👀

File: openhands/sdk/tool/tool.py [hunk 9]
```diff
@@ -290,25 +312,38 @@ def to_openai_tool(
...
+        """
+
+        return {
+            "type": "function",
+            "name": self.name,
+            "description": self.description,
+            "parameters": self._get_tool_schema(
+                add_security_risk_prediction, action_type
+            ),
+            "strict": True,
```
Comment (enyst) [inline openhands/sdk/tool/tool.py]:
```suggestion
            "strict": False,
```

**Comment by enyst** [comment]
@xingyaoww Works for me with `strict=False`!

I think that’s fine?

**Comment by xingyaoww** [comment]
@enyst yeah, that's fine, i was trying to get it to work: https://github.com/All-Hands-AI/agent-sdk/commit/4661859d4f673e773722bfb2169be0bf4611b071

But it is VERY tricky to get MCP schema to work :( In the mean time i discover another bug which is "kind" of the Action Input schema somehow show up in "properties" 😓 (unrelated to this PR)

**Review by xingyaoww — APPROVED** [review]
Confirmed it works locally for me! Let's get this in!

File: openhands/sdk/llm/message.py (hunk 8 above)
Comment (xingyaoww) [reply openhands/sdk/llm/message.py:576]: @OpenHands create a github issue that we should remove all definition and usage of `to_llm_dict` and `from_litellm_message` in favor of the new API.

**Comment by xingyaoww** [comment]
@OpenHands create another github issue that Tool.to_mcp_schema() will outputs a MCP schema and its "properties" contains "kind" which suppose NOT be contains in tool schema we send to LLM. 

We should consider keeping `kind` as a Pydantic field, and just remove if in the `to_mcp_schema` generation of `openhands/sdk/tool/schema.py`. Also, we can have Schema directly inherit from `DiscriminatedUnionMixin` instead of BaseModel, so that it makes sense to just exclude the `kind` field in the same class.

**Comment by simonrosenberg** [comment]
@enyst it seems that gpt5-mini is crashing because of this fix because GPT-5 does NOT natively support Responses API.

**Comment by enyst** [comment]
I suspect Litellm seems to have some wonky criteria where it doesn't recognize "litellm_proxy/openai/gpt-5" as supporting Responses, but it does recognize "openai/gpt-5". So it reroutes the first.

We can look more into it tomorrow! I'm pretty sure GPT-5 does support natively Responses API. The call does go to OpenAI on `/responses` endpoint, and OpenAI documents support, e.g.
https://platform.openai.com/docs/guides/migrate-to-responses#responses-benefits
